### PR TITLE
[#45] Rewrite to Multiple Public Libraries

### LINKED ISSUE #45: Rewrite to Multiple Public Libraries
And update the blog post accordingly.

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #45

This doesn't compile. I see the following parsing error with this executable stanza:

```
Errors encountered when parsing cabal file ./containers-backpack.cabal:

containers-backpack.cabal:142:43: error:
unexpected ':'
expecting "-", white space, "(", "hiding", "requires", comma or end of input

  140 |   build-depends:       containers-backpack:{ int-strict, ordered-strict, unordered-strict, contrib }
  141 | 
  142 |   mixins:              containers-backpack:contrib (Map.Contrib.Group as Map.Contrib.Group.Int)
      |                                           ^
```

Unfortunately, this means that the interface is not usable at the moment until this bug is fixed in `cabal`. There were some other issues, like problems with having two signatures for `readonly` and not only containers, but they are solvable, and this one is show-stopper :disappointed: 

However, you can have an example of how better it will become!

### CODE REVIEW

**Comment by chshersh** [comment]
Opened a bug report in `cabal`:

* https://github.com/haskell/cabal/issues/6281

File: src/readonly/Map.hsig [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+data Map k v
+class Key k
+
+
+
+
+
+
+
```
Comment (vrom911) [inline src/readonly/Map.hsig]: Is this a generated file?

File: src/sig/Map.hsig [hunk 2]
```diff
@@ -19,28 +18,43 @@ signature Map
...
 lookup        :: Key k => k -> Map k v -> Maybe v
 lookupDefault :: Key k => v -> k -> Map k v -> v
 
+insert        :: Key k => k -> v -> Map k v -> Map k v
+insertWith    :: Key k => (v -> v -> v) -> k -> v -> Map k v -> Map k v
+adjust        :: Key k => (a -> a) -> k -> Map k a -> Map k a
+update        :: Key k => (a -> Maybe a) -> k -> Map k a -> Map k a
+delete        :: Key k => k -> Map k v -> Map k v
+alter         :: Key k => (Maybe v -> Maybe v) -> k -> Map k v -> Map k v
+
```
Comment (vrom911) [inline src/sig/Map.hsig]: Could we ask GitHub to highlight `hsig` files as Haskell? :slightly_smiling_face:

File: src/readonly/Map.hsig (hunk 1 above)
Comment (chshersh) [reply src/readonly/Map.hsig]: No, it's not generated. There were two signatures before: for readonly containers and for modifiable containers. But this doesn't work for some reason in this new solution. So I have only one file now, but haven't deleted old files yet. Will cleanup later :+1:

File: src/sig/Map.hsig (hunk 2 above)
Comment (chshersh) [reply src/sig/Map.hsig]: I also would love to highlight them in my editor :slightly_smiling_face: But I don't know how to do this...

**Comment by chshersh** [comment]
@vrom911 Thanks for your review! Answering your questions:

1. It was a multipackage project before. But now there's only a single package. So `cabal.project` is not needed anymore :slightly_smiling_face: 
2. If I understand correctly, you can't specify different versions for different multiple public libraries. So I'm not sure how you can specify different versions of sublibraries. This is a new approach and we have to figure out how to support and maintain such libraries...
3. Yes, this sounds reasonable :+1: Pushing refactorings via a separate PR will make this PR smaller and easier to re-review in future.

**Comment by chshersh** [comment]
@vrom911 This PR is ready for review! It finally builds locally, the executable is working, all tests are passing, benchmarks are running. Some highlights and summary of changes:

1. Now instead of many small packages, we have a single package with multiple public libraries.
2. To depend on a library, you need to write dependency like this:
    ```
    containers-backpack:ord-strict
    ```
    where `containers-backpack` is the name of this package (we can rename it any time, any suggestions are welcome) and `ord-strict` is the name of a public library inside this package.
3. However, sometimes this doesn't work inside the package itself. So I just specify dependency simply as `ord-strict`. This is due to Cabal bugs.
4. I merged readonly signatures and modifiable signatures. They complicated things a lot. I left `primitive-containers` package for now. But probably we should drop support for it eventually. We can't support it completely anyway (it has 5 different Map data types and we support only one at the moment).

Any feedback is appreciated!

File: containers-backpack.cabal [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                       OverloadedStrings
+                       RecordWildCards
+                       ScopedTypeVariables
+                       StandaloneDeriving
+                       TupleSections
+                       TypeApplications
+                       ViewPatterns
+
+-- signatures for containers
+library sig
```
Comment (vrom911) [inline containers-backpack.cabal]: So there is no difference between internal libraries and libraries to export?

File: test/Main.hs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import Test.QuickCheck (Arbitrary, arbitrary, choose, vectorOf)
+
+import qualified Data.Text as T
+import qualified Map.Laws.Hash as Hash
+import qualified Map.Laws.Int as Int
+import qualified Map.Laws.Ord as Ord
+
+
+-- | ShortText is a Text with length [0, 3] chosen "arbitrarily"
+newtype ShortText = ShortText Text
```
Comment (vrom911) [inline test/Main.hs]: Should this be named field `unShortText`?

File: test/Main.hs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    arbitrary = ShortText . T.pack <$> (choose (0, 3) >>= flip vectorOf arbitrary)
+
+type K = Int
+type V = ShortText
+
+main :: IO ()
+main = do
+  Int.checkLaws (Proxy @Int) (Proxy @V)
+  Ord.checkLaws (Proxy @K) (Proxy @V)
+  Hash.checkLaws (Proxy @K) (Proxy @V)
```
Comment (vrom911) [inline test/Main.hs]: Could we fix indentation here?

File: test/Main.hs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    deriving stock (Show, Generic)
+    deriving newtype (Eq, Ord, Hashable)
+
+{- | Generates a Text with length [0, 3]. Choosing a small text speeds up the
+quickCheck tests.
+
+Note: do not make minimum larger than 0, empty texts are often edge cases.
+-}
+instance Arbitrary ShortText where
+    arbitrary = ShortText . T.pack <$> (choose (0, 3) >>= flip vectorOf arbitrary)
```
Comment (vrom911) [inline test/Main.hs]: Could we add the signature if possible?

File: test/Main.hs (hunk 4 above)
Comment (chshersh) [reply test/Main.hs]: @vrom911 Yes, I guess it should be made according to our style guide. This code was written a long time ago...

File: test/Main.hs (hunk 5 above)
Comment (chshersh) [reply test/Main.hs]: Yes! Sorry, too many changes, forgot to fix everything...

File: containers-backpack.cabal (hunk 3 above)
Comment (chshersh) [reply containers-backpack.cabal]: It's my fault, turns out that all these libraries are indeed internal at the moment. To make them public, I needed to add `visibility: public` field. And, again, the documentation doesn't really help here...

https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-field-library-visibility
https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-section-library-library

Comment (vrom911) [reply containers-backpack.cabal]: :disappointed: we need to write more about such things in public so other people at least would be able to find these tricks easier...

File: containers-backpack.cabal [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  import:              common-options
+  visibility:          public
```
Comment (vrom911) [inline containers-backpack.cabal]: :+1:
